lib/active_model/serializer/adapter.rb

```diff
@@ -55,15 +55,97 @@ def include_meta(json)
+          serializers = fragment_serializer(@serializer.object.class.name, klass)
+
+          cached_serializer = serializers[:cached].constantize.new(@serializer.object)
+          non_cached_serializer = serializers[:non_cached].constantize.new(@serializer.object)
+
+          cached_hash = self.class.new(cached_serializer, @options).serializable_hash
+          non_cached_hash = self.class.new(non_cached_serializer, @options).serializable_hash
+
+          if @serializer.root && self.class == ActiveModel::Serializer::Adapter::JsonApi
+            cached_hash_root(cached_hash, non_cached_hash)
+          else
+            non_cached_hash.merge cached_hash
+          end
         else
           yield
         end
       end
+
+      def cached_hash_root(cached_hash, non_cached_hash)
+        @root
+        hash = {}
+
+        core_cached = cached_hash.first
+        core_non_cached = non_cached_hash.first
+        no_root_cache = cached_hash.delete_if {|key, value| key == core_cached[0] }
+        no_root_non_cache = non_cached_hash.delete_if {|key, value| key == core_non_cached[0] }
+
+        if @root
+          hash[@root] = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
+        else
+          hash = (core_cached[1]) ? core_cached[1].merge(core_non_cached[1]) : core_non_cached[1]
+        end
+
+        hash.merge no_root_non_cache.merge no_root_cache
+      end
+
+      def fragment_associations(serializers, associations)
+        associations.each do |association|
+          options = ",#{association[1][:association_options]}" if association[1].include?(:association_options)
+          eval("#{serializers[:non_cached]}.#{association[1][:type].to_s}(:#{association[0]}#{options})")
+        end
+      end
+
+      def cached_attributes_and_association(klass, serializers)
+        cached_attr = (klass._cache_only) ? klass._cache_only : @serializer.attributes.keys.delete_if {|attr| klass._cache_except.include?(attr) }
+        non_cached_attr = @serializer.attributes.keys.delete_if {|attr| cached_attr.include?(attr) }
+        associations = @serializer.each_association
+
+        cached_attr.each do |attr|
+          if @serializer.each_association.keys.include?(attr)
+            associations.delete(attr)
+            serializers[:cached].constantize.send(@serializer.each_association[attr][:type], attr)
+          end
+        end
+
+        cached_attr.each do |attribute|
+          options = @serializer.class._attributes_keys[attribute]
+          options ||= {}
+          serializers[:cached].constantize.attribute(attribute, options)
+        end
+
+        non_cached_attr.each do |attribute|
+          options = @serializer.class._attributes_keys[attribute]
+          options ||= {}
+          serializers[:non_cached].constantize.attribute(attribute, options)
+        end
+        fragment_associations(serializers, associations)
+      end
```

Comment:
Could you extract all of these methods into another class? `Cache` perhaps? e.g

```
def cached_hash_root(cached_hash, non_cached_hash)
  cache.hash_root(cached_hash, non_cached_hash)
end

# ...

private

def cache
  Cache.new(self)
end
```

Otherwise, it's going to be hard to be maintained.